crates/testing/tests/tests_1/upgrade_task_with_proposal.rs

```diff
@@ -0,0 +1,238 @@
+                build_fake_view_with_leaf(leaves[0].clone()),
+            ),
+        ],
+        upgrade_vote_recvs,
+        vec![
+            QuorumProposalRecv(proposals[1].clone(), leaders[1]),
+            QcFormed(either::Left(proposals[2].data.justify_qc.clone())),
+            SendPayloadCommitmentAndMetadata(
+                build_payload_commitment(&quorum_membership, ViewNumber::new(3)),
+                builder_commitment.clone(),
+                TestMetadata,
+                ViewNumber::new(3),
+                null_block::builder_fee(quorum_membership.total_nodes()).unwrap(),
+            ),
+            VidDisperseSend(vid_dispersals[2].clone(), handle.public_key()),
+            ValidatedStateUpdated(
+                proposals[1].data.view_number(),
+                build_fake_view_with_leaf(leaves[1].clone()),
+            ),
+        ],
+    ];
+
+    let mut proposal_script = TaskScript {
+        timeout: Duration::from_millis(35),
+        state: proposal_state,
+        expectations: vec![
+            Expectations {
+                output_asserts: vec![
+                    exact(UpdateHighQc(genesis_cert.clone())),
+                    exact(HighQcUpdated(genesis_cert.clone())),        
+                ],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![
+                    exact(UpdateHighQc(proposals[1].data.justify_qc.clone())),
+                    exact(HighQcUpdated(proposals[1].data.justify_qc.clone())),   
+                ],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![
+                    exact(UpdateHighQc(proposals[2].data.justify_qc.clone())),
+                    exact(HighQcUpdated(proposals[2].data.justify_qc.clone())),
+                    quorum_proposal_send_with_upgrade_certificate::<TestTypes>()
+                ],
+                task_state_asserts: vec![],
+            },
+        ],
+    };
+
+    let mut upgrade_script = TaskScript {
+        timeout: Duration::from_millis(35),
+        state: upgrade_state,
+        expectations: vec![
+            Expectations {
+                output_asserts: vec![],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![upgrade_certificate_formed::<TestTypes>()],
+                task_state_asserts: vec![],
+            },
+            Expectations {
+                output_asserts: vec![],
+                task_state_asserts: vec![],
+            },
+        ],
+    };
+
+    test_scripts![inputs, proposal_script, upgrade_script].await;
```

Comment:
This is totally a confusing thing, so I apologize for that in advance, but this should actually be using `run_test!` and then we'll just need to change all the inputs from `vec` to `random!` or `serial!`. This will automatically check all input permutations, then you can use `Expectations::from_outputs(all_predicates!` instead of the `vec!` thing. The legacy `test_scripts` macro will likely be removed later.

For an example, check out the quorum proposal task.

Reply:
Oh I was using the same pattern in the original upgrade tests, but yeah, I think it’s a good idea to use random ordering!